awx/sso/common.py

```diff
@@ -0,0 +1,171 @@
+                for sub_dict in teams_dict.values():
+                    for role_name in sub_dict:
+                        if sub_dict[role_name] is None:
+                            continue
+                        if role_name not in roles:
+                            roles.append(role_name)
+            model_roles = Team.objects.filter(name__in=team_names).values_list('name', 'organization__name', *roles, named=True)
+
+        for row in model_roles:
+            for role_name in roles:
+                if object_type == 'organization':
+                    desired_state = desired_states.get(row.name, {})
+                else:
+                    desired_state = desired_states.get(row.organization__name, {}).get(row.name, {})
+
+                if desired_state.get(role_name, None) is None:
+                    # The mapping was not defined for this [org/team]/role so we can just pass
+                    continue
+
+                # If somehow the auth adapter knows about an items role but that role is not defined in the DB we are going to print a pretty error
+                # This is your classic safety net that we should never hit; but here you are reading this comment... good luck and Godspeed.
+                role_id = getattr(row, role_name, None)
+                if role_id is None:
+                    logger.error("{} adapter wanted to manage role {} of {} {} but that role is not defined".format(source, role_name, object_type, row.name))
+                    continue
+
+                if desired_state[role_name]:
+                    # The desired state was the user mapped into the object_type, if the user was not mapped in map them in
+                    if role_id not in users_roles:
+                        logger.debug("{} adapter adding user {} to {} {} as {}".format(source, user.username, object_type, row.name, role_name))
+                        user.roles.add(role_id)
+                else:
+                    # The desired state was the user was not mapped into the org, if the user has the permission remove it
+                    if role_id in users_roles:
+                        logger.debug("{} adapter removing user {} permission of {} from {} {}".format(source, user.username, role_name, object_type, row.name))
+                        user.roles.remove(role_id)
+
+
+def create_org_and_teams(org_list, team_map, adapter, can_create=True):
+    #
+    # org_list is a set of organization names
+    # team_map is a dict of {<team_name>: <org name>}
+    #
+    # Move this junk into save of the settings for performance later, there is no need to do that here
+    #    with maybe the exception of someone defining this in settings before the server is started?
+    # ==============================================================================================================
+
+    if not can_create:
+        logger.debug(f"Adapter {adapter} is not allowed to create orgs/teams")
```

Comment:
in what scenario would this condition be hit? saml auto create disabled? If so, shouldn't we make the logging message indicate that incase the user is unaware of the setting or has been changed?

Reply:
The SAML adapter has a setting that says whether or not it its actually allowed to create orgs/teams. I don't think there is an equivalent for LDAP. I don't think we used to do anything before (besides skip the task) if the setting was off for SAML. So this log would just show up in the debug logs as an FYI for an admin.

Reply:
Makes sense but maybe we should include help text to how to enable it?

Reply:
It should already be there:
![image](https://user-images.githubusercontent.com/32551173/214972267-a137e9e5-5469-4607-833a-2fd1c64b08d9.png)

I suppose we could do something like `Adapter {adapter} not not allowed to create orgs/team see documentation.` or somehow try and key off `{adapter}` to get more specific but that seems like overkill to me. What do you think?

Reply:
see docs is probably enough